iree/compiler/Conversion/HLOToHLO/Passes.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Conversion/HLOToHLO/Passes.h"
+
+#include "mlir/IR/Matchers.h"
+#include "mlir/Pass/PassManager.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/rewriters.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+struct ConvertHLOToHLOPass
+    : public PassWrapper<ConvertHLOToHLOPass, FunctionPass> {
+  void runOnFunction() override {
+    MLIRContext *context = &getContext();
+    OwningRewritePatternList greedyPatterns;
+    mhlo::PopulateComplexLoweringPatterns(context, &greedyPatterns);
+    if (failed(applyPatternsAndFoldGreedily(getOperation(), greedyPatterns))) {
+      return signalPassFailure();
+    }
+  }
+};
+}  // namespace
+
+std::unique_ptr<OperationPass<FuncOp>> createHLOToHLOPass() {
+  return std::make_unique<ConvertHLOToHLOPass>();
+}
+
+static PassRegistration<ConvertHLOToHLOPass> pass(
+    "iree-codegen-hlo-to-hlo",
+    "Convert from XLA-HLO ops XLA-HLO ops compatible with codegen");
```

Comment:
I have a few questions from the "hlo to hlo conversion" and "ops compatible with codegen" naming/comments:

1. What is meant by "codegen" here, specifically? IREE, and other projects, have used XLA-HLO ops together with lower level code generation dialects before. What's different now?
2. Are the passes here generally useful, and if so, should they live upstream and/or be always enabled?
3. What is it about arbitrary XLA-HLO that isn't compatible with codegen? What properties do the resulting XLA-HLO have that they didn't have before?
4. Is this "conversion" idempotent? The naming suggests it could be.

Is there some alternate terminology that could be used here that would more precisely describe what these passes are accomplishing? Maybe "canonicalization"?

Reply:
Overall I am down for removing this pass. I really only included it to remain consistent with the other patterns. The HLO to HLO lowering should be part of the SPIRV or LLVM pipelines overall.

Reply:
1. I am definitely basing it on context but my understanding is anything that lowers to explicitly generated code. AKA LLVM / SPIRV lowerings. Overall you can think of this pass as an internal sanitization pass for what HLO our IREE ingestion supports.

2. The contents of the passes are already upstreamed in tensorflow (namely the rewrite patterns). Typically these types of lowerings are upstreamed independent of passes.

3. Some XLA-HLO isn't actually HLO. Technically torch_index_select is not valid hlo, instead its a convenience constructor for gather. In this case it removes complex number types by decomposing into equivalent real valued operations. 

4. Yes, this should be a idempotent conversion. There are similar passes in the tensorflow lowerings for reducing the number of needed TF2XLA lowerings by converting to equivalent TF before feeding into TF2XLA.

Reply:
Thanks for the answers!

I'm still skeptical of the naming, but some alternatives are pretty long... `HLOSanitization`, `HLOToSanitizedHLO`, `HLOToRealHLO`, `ComplexHLOToRealHLO`.

Reply:
Changed to HLOToCompatibleHLO.